Blog post of how we release open source projects (#404)
zalando/zalando.github.io: _posts/2019-05-09-How-We-Release-Code.md

---
title: How we release open source projects
author: Per Ploug
date: 2019-05-09
banner:

>   image: connected.jpg

Reviewer: I couldn't find this image. Is it somewhere in the repo?

Author: yup, its in the banners folder - jekyll picks up automatically

---

> > Zalando currently have over 200 open source projects on Github and each month there are approximately 2 new projects proposed for publishing as open source.

Reviewer 2: "Zalando currently has..."

> Over the years we've refined our release process to ensure that it is easy and transparent to publish projects as open source and that we keep

> > our maintainers accountable for the well-being of the projects which they oversee.

Reviewer 3: """
Zalando has over 200 open source projects on Github, with about two new 
proposed projects a month. Over the years we've refined our process to 
ensure it is easy and transparent to publish projects as open source, 
and keep maintainers accountable for the well-being of their projects.
"""

> This blog post is an attempt at describing how we manage the process of proposing, reviewing and approving projects to become open source code while at the same time ensuring that code we release follow our compliance rules and that the maintainers of the projects are aware of the risks and responsibilities involved.

Reviewer 3: """
This blog post describes how we manage the process of proposing, reviewing 
and approving projects to become open source, while at the same time ensuring 
project code follows our compliance rules, and the maintainers of the projects 
are aware of their responsibilities.
"""

[Our formal release guidelines are available here](https://opensource.zalando.com/docs/releasing/index/)

## Overview

> The process roughly involves 5 steps which takes the project from internal source code, through a review phase to an incubation organisation, which later either results in being graduated into our main organisation or archived as an inactive project due to lack of activity or maintainers: 

Reviewer 3: """
The process involves five steps that take the project from internal source 
code, through a review phase to our incubator, which eventually results 
in the project being graduated into our top level organisation, or archived 
as an inactive project due to lack of activity or maintainers: 
"""

1. An internal project is proposed for release by a Zalando engineer
2. The project proposal is reviewed by the internal open source review group

> 3. If approved, the project is published on the Zalando Incubator on Github

Reviewer: ```suggestion
3. If approved, the project is published on the Zalando Incubator on GitHub
```

4. The project activity and health is monitored by the open source team

> 5. The project graduates from the incubator and into the main Zalando organisation, or the project is decommission and marked as archived.

Reviewer: decommissionED?
```suggestion
5. The project graduates from the incubator and into the main Zalando organisation, or the project is decommissioned and marked as archived.
```

Reviewer: And not sure about "and into" — is it intentional?

Reviewer 3: > And not sure about "and into" — is it intentional?

can be replaced with "to"

> The part where we monitor projects and decide on whether to keep them or archive them will be outlined in greater detail in a later blogpost. 

Reviewer 3: """
How we monitor incubator projects and decide on whether to promote or archive 
them will be detailed detail in a later blogpost. 
"""

## Proposing a new open source project

> First step in getting an internal project published on the Zalando Incubator is to apply for publication by filling out a google form and understanding our requirements [which are available here](https://opensource.zalando.com/docs/releasing/index/)

Reviewer 3: The first step to getting an internal project published on the Zalando Incubator 
is to fill out a google form and confirm understanding our requirements, 
[which are available here](https://opensource.zalando.com/docs/releasing/index/).

> Anyone inside of Zalando can do this and this step serves 2 purposes:

Reviewer 3: Use a numbered bullet list for a numbered lead in

- **To collect information required to publish a project**, such as its current location, who will be maintaining it and the long term plan for maintaining it. 

> - **To set expectations for the maintainers**, such as amount of time needed to maintain the project, sign-off from the developers lead and ensuring the project does not require Zalando-specific dependencies

Reviewer: Dot in the end?
```suggestion
- **To set expectations for the maintainers**, such as amount of time needed to maintain the project, sign-off from the developers lead and ensuring the project does not require Zalando-specific dependencies.
```

> - **To set expectations for the maintainers**, such as amount of time needed to maintain the project, sign-off from the developers lead and ensuring the project does not require Zalando-specific dependencies

Reviewer 3: """
such as amount of time needed to maintain the project, sign-off from the developers' engineering lead and ensuring the project does not require internal Zalando dependencies.
"""

![Project Release Form](/assets/img/os-release-form.png)

> You can see a public version of the [approval form without validation here](https://goo.gl/forms/9C4xlel5DlIK52Xw1).

Reviewer: This forms has some links to the internal docs, e.g. https://rulesofplay.docs.zalando.net/OS/os/ — not accessible in the anonymous mode.

Reviewer: And in the end, there is an answer field without question:

![image](https://user-images.githubusercontent.com/544296/58040056-c24f3f00-7b34-11e9-930a-773db2a0eb4c.png)

Author: yeah - this is to avoid the form being submitted by externals - will remove the links

Questions addressing who have signed-off on the publishing and how many hours the developers have been granted to maintain the project serve as a good way to set expectations, both to the lead who sign off and for the maintainers. To run a sustainable project requires time and resources and we do not expect developers to use their private time, instead granting work time should be part of the conversation. 

> 

Reviewer 3: """
Questions addressing who signed-off on publishing the project and how many 
hours the developers can commit to maintain the project serve as a good way 
to set expectations, both for the lead who appoved and for the maintainers. 
To run a sustainable project requires a commitment and we do not expect 
developers to use their private time—instead ensuring time will be made 
to work on the project should be part of the conversation. 
"""

We also address the need to have basic project health files in place such as Code of conduct, ways for users to get in touch in case of security issues and to reach out to the maintainers - we provide maintainers with a [boiler-plate set of files](https://github.com/zalando-incubator/new-project) for guidance. We do this for 2 reasons:

> 

Reviewer 3: See above comment re numbered lists (consistency).

Reviewer 3: """
We also address the need to have basic project health files in place such as 
a Code of Conduct, ways for users to get in touch in case of security issues, 
features or bugs, by providing maintainers with a [standaerd set of files](https://github.com/zalando-incubator/new-project). We do this for 2 reasons:
"""

> 1. Ownership of code should be visible to other teams inside Zalando, and to potential audits, beyond ownership, these files also communicate how to contribute, how to report security issues and our code of conduct. 

Reviewer: ```suggestion
1. Ownership of code should be visible to other teams inside Zalando, and to potential audits. Beyond ownership, these files also communicate how to contribute, how to report security issues and our code of conduct. 
```

> 2. Communication channels should be public, so maintainers of a project must be approachable by external contributors, we do not ever intent to just throw code over the wall, so having clear ways to reach our maintainers is a central part of taking active ownership of code. 

Reviewer: Perhaps, better to split into few sentences (both points 1 & 2).

```suggestion
2. Communication channels should be public, so maintainers of a project must be approachable by external contributors. We do not ever intent to just throw code over the wall, so having clear ways to reach our maintainers is a central part of taking active ownership of code. 
```

> 2. Communication channels should be public, so maintainers of a project must be approachable by external contributors, we do not ever intent to just throw code over the wall, so having clear ways to reach our maintainers is a central part of taking active ownership of code. 

Reviewer 3: """
Communication channels must be public, so maintainers of a project can be 
approached by external contributors. We want to avoid the _throw code over 
the wall_ antipattern, so having clear ways to reach our maintainers is a 
central part of taking active ownership of code.
"""

## The Open Source Review Group

> When a project is proposed, it is automatically shared on an internal mailing-list which consists of everyone at Zalando currently maintaining an approved project - which is currently a group of around 200 people - this structure allows us to spread the decision making process across many different people and viewpoints. 

Reviewer 3: """
When a project is proposed, it is automatically shared on an internal 
mailing-list that consists of everyone at Zalando currently maintaining 
an approved project. This group is currently about two hundred people, 
which allows us to spread the decision making process across many 
different people and viewpoints. 
"""

![Review Group](/assets/img/review-group.png)

### Discussing the why

> The point here is to have as many eyes on the proposal as possible, specifically we are interested in discussing the WHY of releasing a project and the 3 questions below is central in this discussion:

Reviewer: ```suggestion
The point here is to have as many eyes on the proposal as possible. Specifically, we are interested in discussing the WHY of releasing a project and the 3 questions below are central in this discussion:
```

> The point here is to have as many eyes on the proposal as possible, specifically we are interested in discussing the WHY of releasing a project and the 3 questions below is central in this discussion:

Reviewer 3: """
The point here is to have as many eyes on the proposal as possible. 
Specifically we are interested in discussing the WHY of releasing a 
project and the 3 questions below are central to this discussion:
"""

1. Can the project be sustainable?

> 2. Do Zalando have any value in taking ownership?

Reviewer 3: > Do Zalando have any value in taking ownership?

How is Zalando taking ownership of something it already owns valuable?

Author: Good point - rephrased.

> 3. Does it have any value to anyone outside Zalando?

Reviewer 3: """
1. Will the project be sustainable?
"""

> When source code is released as open source, you are essentially giving away something of value, you are also taking responsibility for spending time and resources on the additional overhead associated with open sourcing, this loss of value and use of resources should be justified, there are multiple ways to look at this, such as:

Reviewer: It also looks like few separate sentences here, not comma-separated parts.

```suggestion
When source code is released as open source, you are essentially giving away something of value, you are also taking responsibility for spending time and resources on the additional overhead associated with open sourcing. This loss of value and use of resources should be justified. There are multiple ways to look at this, such as:
```

> When source code is released as open source, you are essentially giving away something of value, you are also taking responsibility for spending time and resources on the additional overhead associated with open sourcing, this loss of value and use of resources should be justified, there are multiple ways to look at this, such as:

Reviewer 3: """
When code is released as open source, you are essentially sharing 
something of value, and, you are also taking responsibility for 
committing time to the additional overhead associated with open sourcing. 
This commitment and exchange of value should be justified. There are 
multiple ways to look at this, such as:
"""

> 1. The project contributes positively to the employer branding efforts and supports hiring of tech talent

Reviewer 3: This is stated as a requirement, but can't be proven before release. How is this measured and what input and expectations come from EP?

> 2. The project helps establish the company as a leader in a certain domain

Reviewer 3: Why is this a requirement?

Author: Open source requires time and effort, and it's constantly questioned if it makes sense to do. While it is hard to always find justification to open sourcing code on these parameters, it should never be the case, that open source is considered a "special" process where you do not need to  justify it.

3. The project will gain features and bugfixes from external community members
4. The maintainer team could gain valuable knowledge through collaborating with external community members

> For Zalando we've seen several projects contribute to our employer branding efforts, it is however a byproduct and should not be the main reason for open sourcing. It is ofcourse nice that Zalando is recognised for its Kubernetes ([External-Dns](https://github.com/kubernetes-incubator/external-dns), [Stackset-Controller](https://github.com/zalando-incubator/stackset-controller), [es-operator](https://github.com/zalando-incubator/es-operator), PostGreSQL ([Patroni](https://github.com/zalando/patroni) and [postgres-operator](https://github.com/zalando/postgres-operator) and Machine Learning projects ([Flair](https://github.com/zalandoresearch/flair) and [Fashion-mnist](https://github.com/zalandoresearch/fashion-mnist), but it is a value which is hard to measure the effect of and not a long-term motivator for neither the maintainer or company. 

Reviewer 2: "It is of course..."

Reviewer 4: It's PostgreSQL, not PostGreSQL.

> For Zalando we've seen several projects contribute to our employer branding efforts, it is however a byproduct and should not be the main reason for open sourcing. It is ofcourse nice that Zalando is recognised for its Kubernetes ([External-Dns](https://github.com/kubernetes-incubator/external-dns), [Stackset-Controller](https://github.com/zalando-incubator/stackset-controller), [es-operator](https://github.com/zalando-incubator/es-operator), PostGreSQL ([Patroni](https://github.com/zalando/patroni) and [postgres-operator](https://github.com/zalando/postgres-operator) and Machine Learning projects ([Flair](https://github.com/zalandoresearch/flair) and [Fashion-mnist](https://github.com/zalandoresearch/fashion-mnist), but it is a value which is hard to measure the effect of and not a long-term motivator for neither the maintainer or company. 

Reviewer: ```suggestion
For Zalando, we've seen several projects contribute to our employer branding efforts. It is however a byproduct and should not be the main reason for open sourcing. It is of course nice that Zalando is recognised for its Kubernetes ([External-Dns](https://github.com/kubernetes-incubator/external-dns), [Stackset-Controller](https://github.com/zalando-incubator/stackset-controller), [es-operator](https://github.com/zalando-incubator/es-operator), PostGreSQL ([Patroni](https://github.com/zalando/patroni) and [postgres-operator](https://github.com/zalando/postgres-operator) and Machine Learning projects ([Flair](https://github.com/zalandoresearch/flair) and [Fashion-mnist](https://github.com/zalandoresearch/fashion-mnist), but it is a value which is hard to measure the effect of and not a long-term motivator for neither the maintainer or company. 
```

Reviewer: There is also a styling issue with "It is however … It is of course …".

> For Zalando we've seen several projects contribute to our employer branding efforts, it is however a byproduct and should not be the main reason for open sourcing. It is ofcourse nice that Zalando is recognised for its Kubernetes ([External-Dns](https://github.com/kubernetes-incubator/external-dns), [Stackset-Controller](https://github.com/zalando-incubator/stackset-controller), [es-operator](https://github.com/zalando-incubator/es-operator), PostGreSQL ([Patroni](https://github.com/zalando/patroni) and [postgres-operator](https://github.com/zalando/postgres-operator) and Machine Learning projects ([Flair](https://github.com/zalandoresearch/flair) and [Fashion-mnist](https://github.com/zalandoresearch/fashion-mnist), but it is a value which is hard to measure the effect of and not a long-term motivator for neither the maintainer or company. 

Reviewer 3: """
At Zalando we've seen several projects contribute to our employer branding 
efforts, it is however a side-effect and should not be the main reason for 
open sourcing. It is of course nice that Zalando is recognised for its 
Kubernetes ([External-Dns](https://github.com/kubernetes-incubator/external-dns), [Stackset-Controller](https://github.com/zalando-incubator/stackset-controller), [es-operator](https://github.com/zalando-incubator/es-operator), PostGreSQL ([Patroni](https://github.com/zalando/patroni) and [postgres-operator](https://github.com/zalando/postgres-operator), as well as Machine Learning projects such as ([Flair](https://github.com/zalandoresearch/flair) and [Fashion-mnist](https://github.com/zalandoresearch/fashion-mnist). Nonethless it is hard to measure the brand 
impact of such projects, and not a long-term motivation for the maintainers or Zalando. 
"""

> ### Reviwing project quality

Reviewer 2: "Reviewing..."

Besides discussing the WHY, the open source team looks at compliance-specific areas which could be a blocker for releasing:

1. Do we use dependencies which have incompatible licensing
2. Does the source code contain anything confidential (such as tokens, urls, passwords, etc) 

> 3. Does the project contain functionality or IP which gives Zalando a competetive advantage (such as the code that powers our search results)

Reviewer: Maybe it is better to write IP in full? I read it as an IP-address in the first iteration.

```suggestion
3. Does the project contain functionality or intellectual property which gives Zalando a competetive advantage (such as the code that powers our search results)
```

4. Is the project something Zalando would consider trying to patent?

We use a dependency licensing scanning tool, as well as a source code scanner to look for tokens and passwords, to automate this as much as possible.

### Review Meeting

> Once a month the review group sits down with the maintainers proposing new projects, here the dicussions from the mailing-list will be concluded and a release decision is made. The project is either granted release, rejected, or the maintainers are asked to improve certain aspects of the project before it can be released, depending on the number of proposals these decisions are reached in 30-60 minutes and release meetings are setup to formally prepare the files and publish them on Github.com.

Reviewer: ```suggestion
Once a month the review group sits down with the maintainers proposing new projects. Here, the dicussions from the mailing-list will be concluded and a release decision is made. The project is either granted release, rejected, or the maintainers are asked to improve certain aspects of the project before it can be released. Depending on the number of proposals, these decisions are reached in 30-60 minutes and release meetings are set up to formally prepare the files and publish them on Github.com.
```

> Once a month the review group sits down with the maintainers proposing new projects, here the dicussions from the mailing-list will be concluded and a release decision is made. The project is either granted release, rejected, or the maintainers are asked to improve certain aspects of the project before it can be released, depending on the number of proposals these decisions are reached in 30-60 minutes and release meetings are setup to formally prepare the files and publish them on Github.com.

Reviewer 3: """
Once a month the review group sits down with the maintainers proposing new 
projects. The discussions from the mailing-list are considered by the group 
and a decision is made. The project is either released, rejected, or, 
the maintainers are asked to improve certain aspects of the project before 
it can be released. 
"""

> Depending on the number of proposals, these decisions are reached in 30-60 minutes and release meetings are set up to formally prepare the files and publish them on Github.com.

I wasn't able to follow this—understanding what "release meetings" are would probably help.

## Publishing the source code

> Finally, after mailing-list discussions and approval in the monthly meeting, the project is formally released, we have a somewhat specific way of doing this:

Reviewer 3: """
After mailing-list discussion and approval in the monthly meeting, the project 
is released. We have a specific approach to doing this:
"""

> 1. We only transfer the current state of the repository to github, so we do not include the git history, while having the history would be very valuable to track down decisions for code changes, it is simply too big of a security risk and would require the maintainers to audit all commits.

Reviewer: ```suggestion
1. We only transfer the current state of the repository to github, so we do not include the git history. While having the history would be very valuable to track down decisions for code changes, it is simply too big of a security risk and would require the maintainers to audit all commits.
```

> 2. We automatically merge project files with our baseline files, to ensure all repositories have a minimal set of files, these are templated with employee names, emails and github names, so contact info and meta data is consistent. 

Reviewer: ```suggestion
2. We automatically merge project files with our baseline files, to ensure all repositories have a minimal set of files. These files are templated with employee names, emails and github names, so contact info and meta data is consistent. 
```

> 3. The project is setup with a dedicated team assigned to it, with the correct branch protection in place and compliance tooling installed by default (we have a bot called [Zincr](https://github.com/zalando-incubator/zincr) for this)

Reviewer: ```suggestion
3. The project is set up with a dedicated team assigned to it, with the correct branch protection in place and compliance tooling installed by default (we have a bot called [Zincr](https://github.com/zalando-incubator/zincr) for this).
```

> And that is our release process for initially releasing new projects, I hope it gave you an insight into what a company of Zalando's size have to consider before releasing new code and how we have tried to keep the process simple and transparent for the maintainers of our projects. 

Reviewer: ```suggestion
And that is our release process for initially releasing new projects. I hope it gave you an insight into what a company of Zalando's size have to consider before releasing new code and how we have tried to keep the process simple and transparent for the maintainers of our projects. 
```

Reviewer 3: >have to consider 

"has to consider"
